salt/_states/better_kubernetes.py

```diff
@@ -0,0 +1,87 @@
+"""Management of Kubernetes objects as Salt states.
+
+This module defines two state functions: `object_present` and `object_absent`.
+Those will then simply delegate all the logic to the `metalk8s_kubernetes`
+execution module, only managing simple dicts in this state module.
+"""
+from salt.exceptions import CommandExecutionError
+
+__virtualname__ = 'metalk8s_kubernetes'
+
+
+def __virtual__():
+    if 'metalk8s_kubernetes.ping' not in __salt__:
+        return False, '`metalk8s_kubernetes` module is unavailable'
+    return __virtualname__
+
+def object_absent(name, kubeconfig, context, manifest):
+    """Ensure that the object is absent."""
+    ret = {'name': name, 'changes': {}, 'result': False, 'comment': ''}
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name, kubeconfig, context, manifest
+    )
+
+    if obj is None:
+        ret['result'] = True if not __opts__['test'] else None
+        ret['comment'] = 'The object does not exist'
+        return ret
+
+    if __opts__['test']:
+        ret['result'] = None
+        ret['comment'] = 'The object is going to be deleted'
+        return ret
+
+    try:
+        __salt__['metalk8s_kubernetes.delete_object'](obj)
+    except CommandExecutionError as exc:
+        ret['comment'] = 'Something went wrong: {}'.format(exc)
+    else:
+        ret['result'] = True
+        ret['changes'] = {'old': 'present', 'new': 'absent'}
+        ret['comment'] = 'The object was deleted'
+
+    return ret
+
+
+def object_present(name, kubeconfig, context, manifest):
+    """Ensure that the object is present."""
+    ret = {'name': name, 'changes': {}, 'result': False, 'comment': ''}
+
+    obj = __salt__['metalk8s_kubernetes.get_object'](
+        name, kubeconfig, context, manifest
+    )
```

Comment:
IMO we really want to handle idempotancy

Reply:
You mean avoid replacing the object if it matches the desired manifest? I'd love to, and will give it a try once the rest works fine.

However, from discussions with @NicolasT, we anticipated complications:
- how do we handle default values for fields unset in the source manifest (and what would be the difference between a default value and an unset field, from the client perspective)?
- how do we know which fields are not supposed to be managed by the client (e.g. `status`, `metadata.creationTimestamp`, ...)?

Those two issues have simple solutions that contradict each other:
- if we only check fields set in the source manifest, we can't ensure other fields were not set manually by a user, and that's undesired behaviour
- if we check all fields except a known set of exceptions, we may replace greedily and lose idempotency for all default values that were unset in the source manifest

I'd settle for the second approach, as it's still better than not even trying to be idempotent. But that wouldn't be idempotent most of the time, since **a lot** of fields in standard objects are optional. Maybe, though that would require some extensive investigations, we could ignore fields set to `None`, but I'm not convinced that it would map to unset optional fields directly.